### PR TITLE
Custom resource encryption

### LINKED ISSUE #104165: Custom resources don't get encrypted in etcd
#### What happened:
Trying to encrypt a custom resource in etcd with the following EncryptionConfiguration in kube-apiserver does not lead to encrypted state in etcd.

In this example the custom resource is named "tokens" and belongs to the group "management.cattle.io":

`apiVersion: apiserver.config.k8s.io/v1
kind: EncryptionConfiguration
resources:
- providers:
  - aescbc:
      keys:
      - name: some-key
        secret: some-secret
  - identity: {}
  resources:
  - tokens.management.cattle.io
- providers:
  - aescbc:
      keys:
      - name: some-key
        secret: some-secret
  - identity: {}
  resources:
  - secrets`

#### What you expected to happen:
Expecting custom resources to be encrypted in etcd if queries e.g. like this:
`etcdctl get /registry/management.cattle.io/tokens/some-token`

#### How to reproduce it (as minimally and precisely as possible):

Activate encryption on the kube-apiserver with a `EncryptionConfiguration` definition on any custom resource and query state in etcd.

#### Anything else we need to know?:
Tested this both with namespaced and non-namespaced custom resources. In both cases encryption does not work.

#### Environment:
- Kubernetes version: v1.20.6
- Cloud provider or hardware configuration: Openstack
- OS: flatcar 2765.2.3
- Kernel: Linux host-10-13-39-90 5.10.32-flatcar #1 SMP Tue Apr 27 22:38:30 -00 2021 x86_64 Intel Core Processor (Haswell, no TSX, IBRS) GenuineIntel GNU/Linux
- Install tools: rke version v1.2.9

### PR DESCRIPTION
**Pull request by ritazh**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
Enable encryption of custom resources in ETCD

#### Which issue(s) this PR fixes: 

Fixes #104165

#### Does this PR introduce a user-facing change?

```release-note
kube-apiserver: APIs backed by CustomResourceDefinitions can now be specified in the `--encryption-provider-config` file; when specified, the data for those custom resources can now be encrypted in etcd.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
Custom resources can now be encrypted in ETCD
```

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/options/options.go [hunk 1]
```diff
@@ -43,9 +45,10 @@ const defaultEtcdPathPrefix = "/registry/apiextensions.kubernetes.io"
 
 // CustomResourceDefinitionsServerOptions describes the runtime options of an apiextensions-apiserver.
 type CustomResourceDefinitionsServerOptions struct {
-	ServerRunOptions   *genericoptions.ServerRunOptions
-	RecommendedOptions *genericoptions.RecommendedOptions
-	APIEnablement      *genericoptions.APIEnablementOptions
+	ServerRunOptions     *genericoptions.ServerRunOptions
+	RecommendedOptions   *genericoptions.RecommendedOptions
+	APIEnablement        *genericoptions.APIEnablementOptions
+	TransformerOverrides map[schema.GroupResource]value.Transformer
```
Comment (enj) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/options/options.go:51]: `CustomResourceDefinitionsServerOptions` is not a struct that this PR should be changing.

File: test/integration/controlplane/transformation/transformation_testcase.go [hunk 2]
```diff
@@ -134,17 +139,39 @@ func (e *transformTest) run(unSealSecretFunc unSealSecret, expectedEnvelopePrefi
...
+			e.logger.Errorf("expected %s from KubeAPI, but got %s", secretVal, string(s.Data[secretKey]))
+		}
+	} else {
+		e.logger.Logf("Get object with dynamic client")
+		dynamicClient, err := dynamic.NewForConfig(e.kubeAPIServer.ClientConfig)
+		if err != nil {
+			e.logger.Fatalf("Unexpected error: %v", err)
+		}
+		fooResource := schema.GroupVersionResource{Group: e.group, Version: "v1", Resource: e.kind}
+		obj, err := dynamicClient.Resource(fooResource).Get(context.TODO(), e.name, metav1.GetOptions{})
```
Comment (enj) [inline test/integration/controlplane/transformation/transformation_testcase.go:168]: This does not look like it handles namespaces correctly?

File: test/integration/controlplane/transformation/transformation_testcase.go [hunk 3]
```diff
@@ -167,6 +201,7 @@ func (e *transformTest) getRawSecretFromETCD() ([]byte, error) {
 	if err != nil {
 		return nil, fmt.Errorf("failed to read %s from etcd: %v", secretETCDPath, err)
 	}
+	e.logger.Logf("path: %s", secretETCDPath)
```
Comment (enj) [inline test/integration/controlplane/transformation/transformation_testcase.go:204]: I would rename the functions and variables to be more generic:

```suggestion
	e.logger.Logf("path: %s", etcdPath)
```

File: test/integration/controlplane/transformation/transformation_testcase.go [hunk 4]
```diff
@@ -134,17 +139,39 @@ func (e *transformTest) run(unSealSecretFunc unSealSecret, expectedEnvelopePrefi
...
+			e.logger.Fatalf("Unexpected error: %v", err)
+		}
+		fooResource := schema.GroupVersionResource{Group: e.group, Version: "v1", Resource: e.kind}
+		obj, err := dynamicClient.Resource(fooResource).Get(context.TODO(), e.name, metav1.GetOptions{})
+		if err != nil {
+			e.logger.Errorf("Failed to get test instance: %v, name: %s", err, e.name)
+		}
+		if obj.GetName() != e.name {
+			e.logger.Errorf("expected %s from KubeAPI, but got %s", e.name, obj.GetName())
+		}
```
Comment (enj) [inline test/integration/controlplane/transformation/transformation_testcase.go:174]: Check that the GVR + namespace + name match what you expect (since those 5 things uniquely identify an object).

File: test/integration/controlplane/transformation/transformation_testcase.go [hunk 5]
```diff
@@ -134,17 +139,39 @@ func (e *transformTest) run(unSealSecretFunc unSealSecret, expectedEnvelopePrefi
...
+	} else {
+		e.logger.Logf("Get object with dynamic client")
+		dynamicClient, err := dynamic.NewForConfig(e.kubeAPIServer.ClientConfig)
+		if err != nil {
+			e.logger.Fatalf("Unexpected error: %v", err)
+		}
+		fooResource := schema.GroupVersionResource{Group: e.group, Version: "v1", Resource: e.kind}
+		obj, err := dynamicClient.Resource(fooResource).Get(context.TODO(), e.name, metav1.GetOptions{})
+		if err != nil {
+			e.logger.Errorf("Failed to get test instance: %v, name: %s", err, e.name)
```
Comment (enj) [inline test/integration/controlplane/transformation/transformation_testcase.go:170]:
```suggestion
			e.logger.Fatalf("Failed to get test instance: %v, name: %s", err, e.name)
```

File: test/integration/controlplane/transformation/kms_transformation_test.go [hunk 6]
```diff
@@ -141,7 +141,10 @@ resources:
 	if err != nil {
 		t.Fatalf("Failed to create test secret, error: %v", err)
 	}
-
+	test.group = ""
+	test.kind = "secrets"
+	test.name = testSecret
+	test.ns.Name = testNamespace
```
[the remark this answers is not in the document]
Comment (ritazh) [reply test/integration/controlplane/transformation/kms_transformation_test.go:147]: I made `run` more generic unlike now it always expects the test resource to be a secret. For that, I had to update existing tests to provide the group, kind, name, and namespace as variables.

File: test/integration/controlplane/transformation/all_transformation_test.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+		_, err = dynamicClient.Create(ctx, obj, metav1.CreateOptions{})
+		if err != nil {
+			t.Fatal(err)
+		}
+		if _, err := dynamicClient.Get(ctx, obj.GetName(), metav1.GetOptions{}); err != nil {
+			t.Fatalf("object should exist: %v", err)
+		}
+	}
+}
+func TestEncryptSupportedForAllResourceTypes(t *testing.T) {
```
Comment (enj) [inline test/integration/controlplane/transformation/all_transformation_test.go:75]: I confirmed that this test fails against current `HEAD` as expected:

```
=== CONT  TestEncryptSupportedForAllResourceTypes
    transformation_testcase.go:135: expected data to be prefixed with k8s:enc:aescbc:v1:key1:, but got {"apiVersion":"awesome.bears.com/v1","kind":"Panda","metadata":{"creationTimestamp":"2022-10-17T17:28:37Z","generation":1,"managedFields":[{"apiVersion":"awesome.bears.com/v1","fieldsType":"FieldsV1","fieldsV1":{"f:spec":{".":{},"f:replicas":{}}},"manager":"transformation.test","operation":"Update","time":"2022-10-17T17:28:37Z"}],"name":"cr3panda","uid":"c3e46d3b-94bf-4302-af08-47284045dc74"},"spec":{"replicas":100}}
```

**Comment by enj** [comment]
New integration test from 0c2357af6aa40ec0f6b5cae1fc60c748eec997a1 applied to `master` fails as expected, with proper subtest breakdown:

```
=== RUN   TestEncryptSupportedForAllResourceTypes
    testserver.go:244: runtime-config=map[api/all:true]
    testserver.go:245: Starting kube-apiserver on port 36789...
    testserver.go:265: Waiting for /healthz to be ok...
=== RUN   TestEncryptSupportedForAllResourceTypes/configmaps
=== PAUSE TestEncryptSupportedForAllResourceTypes/configmaps
=== RUN   TestEncryptSupportedForAllResourceTypes/customresourcedefinitions
=== PAUSE TestEncryptSupportedForAllResourceTypes/customresourcedefinitions
=== RUN   TestEncryptSupportedForAllResourceTypes/pandas
=== PAUSE TestEncryptSupportedForAllResourceTypes/pandas
=== RUN   TestEncryptSupportedForAllResourceTypes/apiservices
=== PAUSE TestEncryptSupportedForAllResourceTypes/apiservices
=== RUN   TestEncryptSupportedForAllResourceTypes/pods
=== PAUSE TestEncryptSupportedForAllResourceTypes/pods
=== CONT  TestEncryptSupportedForAllResourceTypes/configmaps
=== CONT  TestEncryptSupportedForAllResourceTypes/apiservices
=== CONT  TestEncryptSupportedForAllResourceTypes/pandas
=== CONT  TestEncryptSupportedForAllResourceTypes/customresourcedefinitions
=== CONT  TestEncryptSupportedForAllResourceTypes/pods
=== CONT  TestEncryptSupportedForAllResourceTypes/customresourcedefinitions
    transformation_testcase.go:194: Get object with dynamic client
=== CONT  TestEncryptSupportedForAllResourceTypes/apiservices
    transformation_testcase.go:194: Get object with dynamic client
=== CONT  TestEncryptSupportedForAllResourceTypes/pandas
    transformation_testcase.go:135: expected data to be prefixed with k8s:enc:aescbc:v1:key1:, but got {"apiVersion":"awesome.bears.com/v1","kind":"Panda","metadata":{"creationTimestamp":"2022-10-18T12:39:44Z","generation":1,"managedFields":[{"apiVersion":"awesome.bears.com/v1","fieldsType":"FieldsV1","fieldsV1":{"f:spec":{".":{},"f:replicas":{}}},"manager":"transformation.test","operation":"Update","time":"2022-10-18T12:39:44Z"}],"name":"cr3panda","uid":"9e9d63b3-d206-4905-b539-cda35cc44f5b"},"spec":{"replicas":100}}
--- FAIL: TestEncryptSupportedForAllResourceTypes (6.39s)
    --- PASS: TestEncryptSupportedForAllResourceTypes/configmaps (0.03s)
    --- PASS: TestEncryptSupportedForAllResourceTypes/customresourcedefinitions (0.03s)
    --- PASS: TestEncryptSupportedForAllResourceTypes/pods (0.03s)
    --- PASS: TestEncryptSupportedForAllResourceTypes/apiservices (0.03s)
    --- FAIL: TestEncryptSupportedForAllResourceTypes/pandas (2.06s)
FAIL
FAIL	k8s.io/kubernetes/test/integration/controlplane/transformation	8.627s
FAIL
```

File: test/integration/controlplane/transformation/all_transformation_test.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+	// - API services from aggregator
+	encryptionConfig := `
+kind: EncryptionConfiguration
+apiVersion: apiserver.config.k8s.io/v1
+resources:
+- resources:
+  - pods
+  - configmaps
+  - customresourcedefinitions.apiextensions.k8s.io
+  - pandas.awesome.bears.com
```
Comment (aramase) [inline test/integration/controlplane/transformation/all_transformation_test.go:88]: 🐼

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/options/options.go [hunk 9]
```diff
@@ -119,20 +123,29 @@ func (o CustomResourceDefinitionsServerOptions) Config() (*apiserver.Config, err
...
+func NewCRDRESTOptionsGetter(etcdOptions genericoptions.EtcdOptions) (genericregistry.RESTOptionsGetter, error) {
+	storageConfig := etcdOptions.StorageConfig
+	storageConfig.Codec = unstructured.UnstructuredJSONScheme
+	etcdOptions.StorageConfig = storageConfig
+
+	etcdOptions.WatchCacheSizes = nil      // this control is not provided for custom resources
+	etcdOptions.SkipHealthEndpoints = true // avoid double wiring of health checks
 
-	return ret
+	c := genericapiserver.Config{}
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/options/options.go:135]: this is... clever... worth a comment explaining what we're doing

File: test/integration/controlplane/transformation/transformation_testcase.go [hunk 10]
```diff
@@ -172,7 +244,9 @@ func (e *transformTest) getRawSecretFromETCD() ([]byte, error) {
 
 func (e *transformTest) getEncryptionOptions() []string {
 	if e.transformerConfig != "" {
-		return []string{"--encryption-provider-config", path.Join(e.configDir, encryptionConfigFileName)}
+		return []string{
+			"--encryption-provider-config", path.Join(e.configDir, encryptionConfigFileName),
+			"--disable-admission-plugins", "ServiceAccount"}
```
Comment (liggitt) [inline test/integration/controlplane/transformation/transformation_testcase.go:249]: we need to disable this because we're creating pods?

File: test/integration/controlplane/transformation/transformation_testcase.go [hunk 11]
```diff
@@ -161,6 +216,23 @@ func (e *transformTest) getETCDPath() string {
 	return fmt.Sprintf("/%s/secrets/%s/%s", e.storageConfig.Prefix, e.ns.Name, e.secret.Name)
 }
 
+func (e *transformTest) getETCDPathForResource(group, resource, name, namespaceName string) string {
+	if e.secret != nil {
```
Comment (liggitt) [inline test/integration/controlplane/transformation/transformation_testcase.go:220]: it's pretty hard to follow for state inside `transformTest` to override params... can we make this a pure function that takes `storagePrefix, group, resource, name, namespace` params?

Comment (liggitt) [reply test/integration/controlplane/transformation/transformation_testcase.go:220]: also, what's the difference between getETCDPathForResource and getETCDPath? can we collapse to a single etcd path computation function?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 12]
```diff
@@ -1134,33 +1131,6 @@ func (d unstructuredDefaulter) Default(in runtime.Object) {
...
-	DefaultWatchCacheSize     int
-	EnableGarbageCollection   bool
-	DeleteCollectionWorkers   int
-	CountMetricPollPeriod     time.Duration
-	StorageObjectCountTracker flowcontrolrequest.StorageObjectCountTracker
-}
-
-func (t CRDRESTOptionsGetter) GetRESTOptions(resource schema.GroupResource) (generic.RESTOptions, error) {
-	ret := generic.RESTOptions{
-		StorageConfig:             t.StorageConfig.ForResource(resource),
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:1150]: verified this matches ApplyTo → StorageFactoryRestOptionsFactory#GetRESTOptions → SimpleStorageFactory#NewConfig

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 13]
```diff
@@ -1134,33 +1131,6 @@ func (d unstructuredDefaulter) Default(in runtime.Object) {
...
-		Decorator:                 generic.UndecoratedStorage,
-		EnableGarbageCollection:   t.EnableGarbageCollection,
-		DeleteCollectionWorkers:   t.DeleteCollectionWorkers,
-		ResourcePrefix:            resource.Group + "/" + resource.Resource,
-		CountMetricPollPeriod:     t.CountMetricPollPeriod,
-		StorageObjectCountTracker: t.StorageObjectCountTracker,
-	}
-	if t.EnableWatchCache {
-		ret.Decorator = genericregistry.StorageWithCacher()
-	}
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:1160]: verified this matches ApplyTo → StorageFactoryRestOptionsFactory#GetRESTOptions when WatchCacheSizes is unset

a v(3) log line will be emitted each time GetRESTOptions is called for a custom resource (on startup and for each CRD creation or modification)

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 14]
```diff
@@ -1134,33 +1131,6 @@ func (d unstructuredDefaulter) Default(in runtime.Object) {
...
-
-func (t CRDRESTOptionsGetter) GetRESTOptions(resource schema.GroupResource) (generic.RESTOptions, error) {
-	ret := generic.RESTOptions{
-		StorageConfig:             t.StorageConfig.ForResource(resource),
-		Decorator:                 generic.UndecoratedStorage,
-		EnableGarbageCollection:   t.EnableGarbageCollection,
-		DeleteCollectionWorkers:   t.DeleteCollectionWorkers,
-		ResourcePrefix:            resource.Group + "/" + resource.Resource,
-		CountMetricPollPeriod:     t.CountMetricPollPeriod,
-		StorageObjectCountTracker: t.StorageObjectCountTracker,
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:1156]: verified EnableGarbageCollection, DeleteCollectionWorkers, CountMetricPollPeriod, StorageObjectCountTracker matches ApplyTo → StorageFactoryRestOptionsFactory#GetRESTOptions

verified ResourcePrefix matches ApplyTo → StorageFactoryRestOptionsFactory#GetRESTOptions → SimpleStorageFactory#ResourcePrefix (and the etcd data integration test confirms the storage path for custom resources is unchanged)

File: test/integration/controlplane/transformation/all_transformation_test.go [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+  - apiservices.apiregistration.k8s.io
```
Comment (liggitt) [inline test/integration/controlplane/transformation/all_transformation_test.go:89]: doesn't need to be in this PR, but we should *definitely* update the godoc on the encryption config API... we don't document at all how to specify group/resources to be encrypted:

```go
	// resources is a list of kubernetes resources which have to be encrypted.
	Resources []string `json:"resources"`
```

File: test/integration/controlplane/transformation/transformation_testcase.go (hunk 10 above)
Comment (ritazh) [reply test/integration/controlplane/transformation/transformation_testcase.go:249]: yes, otherwise the test will end in an error caused by the missing service account:
> error while writing pod: pods "pod1" is forbidden: error looking up service account secret-encryption-test/default: serviceaccount "default"

File: test/integration/controlplane/transformation/transformation_testcase.go (hunk 11 above)
Comment (ritazh) [reply test/integration/controlplane/transformation/transformation_testcase.go:220]: > also, what's the difference between getETCDPathForResource and getETCDPath? can we collapse to a single etcd path computation function?

I didn't want to touch getETCDPath to minimize changes on existing test files. If you prefer to combine them, then I will go ahead and update kms_transformation_test.go and kmsv2_transformation_test.go. LMK!

File: test/integration/controlplane/transformation/all_transformation_test.go (hunk 15 above)
Comment (ritazh) [reply test/integration/controlplane/transformation/all_transformation_test.go:89]: opened https://github.com/kubernetes/kubernetes/issues/113272 to track

File: test/integration/controlplane/transformation/transformation_testcase.go (hunk 11 above)
Comment (liggitt) [reply test/integration/controlplane/transformation/transformation_testcase.go:220]: the ripples don't look *too* bad and I think leave the test package in a less confusing state (mostly just makes the paths that assume they are dealing with secrets explicitly say so)

I got a clean run with this diff:

```diff
diff --git a/test/integration/controlplane/transformation/kms_transformation_test.go b/test/integration/controlplane/transformation/kms_transformation_test.go
index 9773296ee29..3db46c22c3b 100644
--- a/test/integration/controlplane/transformation/kms_transformation_test.go
+++ b/test/integration/controlplane/transformation/kms_transformation_test.go
@@ -145,7 +145,7 @@ resources:
 	// Since Data Encryption Key (DEK) is randomly generated (per encryption operation), we need to ask KMS Mock for it.
 	plainTextDEK := pluginMock.LastEncryptRequest()
 
-	secretETCDPath := test.getETCDPath()
+	secretETCDPath := test.getETCDPathForResource("", "secrets", test.secret.Name, test.secret.Namespace)
 	rawEnvelope, err := test.getRawSecretFromETCD()
 	if err != nil {
 		t.Fatalf("failed to read %s from etcd: %v", secretETCDPath, err)
diff --git a/test/integration/controlplane/transformation/kmsv2_transformation_test.go b/test/integration/controlplane/transformation/kmsv2_transformation_test.go
index dfb250a938f..f8b385adb24 100644
--- a/test/integration/controlplane/transformation/kmsv2_transformation_test.go
+++ b/test/integration/controlplane/transformation/kmsv2_transformation_test.go
@@ -154,7 +154,7 @@ resources:
 	// Since Data Encryption Key (DEK) is randomly generated (per encryption operation), we need to ask KMS Mock for it.
 	plainTextDEK := pluginMock.LastEncryptRequest()
 
-	secretETCDPath := test.getETCDPath()
+	secretETCDPath := test.getETCDPathForResource("", "secrets", test.secret.Name, test.secret.Namespace)
 	rawEnvelope, err := test.getRawSecretFromETCD()
 	if err != nil {
 		t.Fatalf("failed to read %s from etcd: %v", secretETCDPath, err)
diff --git a/test/integration/controlplane/transformation/secrets_transformation_test.go b/test/integration/controlplane/transformation/secrets_transformation_test.go
index b899237a937..ccbd86ef0ff 100644
--- a/test/integration/controlplane/transformation/secrets_transformation_test.go
+++ b/test/integration/controlplane/transformation/secrets_transformation_test.go
@@ -95,7 +95,7 @@ func TestSecretsShouldBeTransformed(t *testing.T) {
 		if err != nil {
 			t.Fatalf("Failed to create test secret, error: %v", err)
 		}
-		test.run(tt.unSealFunc, tt.transformerPrefix)
+		test.runResource(test.logger, tt.unSealFunc, tt.transformerPrefix, "", "v1", "secrets", test.secret.Name, test.secret.Namespace)
 		test.cleanUp()
 	}
 }
diff --git a/test/integration/controlplane/transformation/transformation_testcase.go b/test/integration/controlplane/transformation/transformation_testcase.go
index d5ca2b8fcc7..cea83135866 100644
--- a/test/integration/controlplane/transformation/transformation_testcase.go
+++ b/test/integration/controlplane/transformation/transformation_testcase.go
@@ -114,10 +114,6 @@ func (e *transformTest) cleanUp() {
 	e.kubeAPIServer.TearDownFn()
 }
 
-func (e *transformTest) run(unSealSecretFunc unSealSecret, expectedEnvelopePrefix string) {
-	e.runResource(e.logger, unSealSecretFunc, expectedEnvelopePrefix, "", "v1", "secrets", "", "")
-}
-
 func (e *transformTest) runResource(l kubeapiservertesting.Logger, unSealSecretFunc unSealSecret, expectedEnvelopePrefix,
 	group,
 	version,
@@ -212,17 +208,7 @@ func (e *transformTest) benchmark(b *testing.B) {
 	}
 }
 
-func (e *transformTest) getETCDPath() string {
-	return fmt.Sprintf("/%s/secrets/%s/%s", e.storageConfig.Prefix, e.ns.Name, e.secret.Name)
-}
-
 func (e *transformTest) getETCDPathForResource(group, resource, name, namespaceName string) string {
-	if e.secret != nil {
-		group = ""
-		resource = "secrets"
-		name = testSecret
-		namespaceName = testNamespace
-	}
 	groupResource := resource
 	if group != "" {
 		groupResource = fmt.Sprintf("%s/%s", group, resource)
@@ -234,7 +220,7 @@ func (e *transformTest) getETCDPathForResource(group, resource, name, namespaceN
 }
 
 func (e *transformTest) getRawSecretFromETCD() ([]byte, error) {
-	secretETCDPath := e.getETCDPath()
+	secretETCDPath := e.getETCDPathForResource("", "secrets", e.secret.Name, e.ns.Name)
 	etcdResponse, err := e.readRawRecordFromETCD(secretETCDPath)
 	if err != nil {
 		return nil, fmt.Errorf("failed to read %s from etcd: %v", secretETCDPath, err)
```

**Comment by liggitt** [comment]
/approve
squash, then lgtm

**Comment by sftim** [comment]
My suggestion for the changelog text (which can be Markdown):
```markdown
kube-apiserver: APIs backed by CustomResourceDefinitions can now be specified in the `--encryption-provider-config` file; when specified, the data for those custom resources can now be encrypted in etcd.
```

I like to put command line arguments inside backticks.

**Comment by ritazh** [comment]
Thanks for the suggestion @sftim! I've updated above.
